btree/src/flatfile.rs

```diff
@@ -34,71 +41,103 @@ impl Appender {
         Ok(pos <= MAX_POS_OFFSET)
     }
 
     /// Append a blob of data and return the file offset
     ///
     /// Can only append data of MAX_BLOB_SIZE
-    pub fn append(&mut self, buf: &[u8]) -> Result<Pos, io::Error> {
+    pub fn append(&self, buf: &[u8]) -> Result<Pos, io::Error> {
         if buf.len() > MAX_BLOB_SIZE {
             return Err(Error::new(ErrorKind::Other, "blob size too big"));
         }
-        // if (buf.len() & 0b11) != 0 {
-        //     return Err(Error::new(
-        //         ErrorKind::Other,
-        //         "blob size is not a multiple of 4",
-        //     ));
-        // }
-        let pos = self.ahandle.seek(SeekFrom::Current(0))?;
-        if pos > MAX_POS_OFFSET {
+        if (buf.len() & 0b11) != 0 {
+            return Err(Error::new(
+                ErrorKind::Other,
+                "blob size is not a multiple of 4",
+            ));
+        }
+
+        let next_pos = self.next_pos.load(Ordering::Acquire);
+        let mut storage = self.storage.upgradable_read();
+
+        if next_pos > MAX_POS_OFFSET {
             return Err(Error::new(ErrorKind::Other, "offset position too big"));
         }
+
         let blen = buf.len() as u32;
         let szbuf = blen.to_le_bytes();
-        self.ahandle.write_all(&szbuf[..])?;
-        self.ahandle.write_all(buf)?;
+
+        let region_len = szbuf.len() as u64 + buf.len() as u64;
+
+        let mmaped_region = unsafe {
+            match storage.get_mut(next_pos, region_len) {
+                Ok(slice) => slice,
+                Err(including) => {
+                    {
+                        let mut new_guard = RwLockUpgradableReadGuard::upgrade(storage);
+                        new_guard.extend(including)?;
+                        // the upgradable part here is only so we can assign to the storage variable again
+                        // we won't upgrade again
+                        storage = RwLockWriteGuard::downgrade_to_upgradable(new_guard);
+                    }
+                    storage.get_mut(next_pos, region_len).unwrap()
+                }
+            }
+        };
+
+        self.next_pos
+            .store(next_pos + region_len, Ordering::Release);
+
+        mmaped_region[0..szbuf.len()].copy_from_slice(&szbuf[..]);
+        mmaped_region[szbuf.len()..].copy_from_slice(&buf[..]);
 
         // self.ahandle.sync_data()?;
-        Ok(Pos(pos))
+        Ok(Pos(next_pos))
     }
 
     /// Get the blob stored at position @pos
-    pub fn get_at(&mut self, pos: Pos) -> Result<Box<[u8]>, io::Error> {
-        self.rhandle.seek(SeekFrom::Start(pos.0))?;
+    pub fn get_at(&self, pos: Pos) -> Result<Box<[u8]>, io::Error> {
+        // TODO: this will panic if position if out of range, we may want to handle that?
+        let storage = self.storage.read();
+        let szbuf = unsafe { storage.get(pos.into(), 4) };
+
+        let len = LittleEndian::read_u32(&szbuf);
```

Comment:
std now have `u32::from_le_bytes()` with a `try_into().unwrap()` to copy from slice to array